ICE: `assertion failed: self.is_struct() || self.is_union()`

<!--
Thank you for finding an Internal Compiler Error! 🧊  If possible, try to provide
a minimal verifiable example. You can read "Rust Bug Minimization Patterns" for
how to create smaller examples.
http://blog.pnkfx.org/blog/2019/11/18/rust-bug-minimization-patterns/
-->

### Code

```Rust
#![feature(repr_simd)]

use std::arch::asm;

#[repr(simd)]
enum Es {}
static CLs: Es;

pub fn main() {
    let mut b = 4i32;
    unsafe {
        asm!("{}", out(reg) CLs);
    }
}
```

A mutant of [`tests/pretty/asm.rs`](https://github.com/rust-lang/rust/blob/c90bcb9571b7aab0d8beaa2ce8a998ffaf079d38/tests/pretty/asm.rs)


### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.93.0-nightly (c90bcb957 2025-11-06)
binary: rustc
commit-hash: c90bcb9571b7aab0d8beaa2ce8a998ffaf079d38
commit-date: 2025-11-06
host: x86_64-apple-darwin
release: 1.93.0-nightly
LLVM version: 21.1.3
```

### Error output

Command: `rustc`

```
error: free static item without body
 --> 14.rs:7:1
  |
7 | static CLs: Es;
  | ^^^^^^^^^^^^^^-
  |               |
  |               help: provide a definition for the static: `= <expr>;`

error[E0517]: attribute should be applied to a struct
 --> 14.rs:5:8
  |
5 | #[repr(simd)]
  |        ^^^^
6 | enum Es {}
  | ---------- not a struct

error[E0084]: unsupported representation for zero-variant enum
 --> 14.rs:5:8
  |
5 | #[repr(simd)]
  |        ^^^^
6 | enum Es {}
  | ------- zero-variant enum


thread 'rustc' (6698172) panicked at compiler/rustc_middle/src/ty/adt.rs:462:9:
assertion failed: self.is_struct() || self.is_union()
```

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Backtrace</strong></summary>
<p>

```
thread 'rustc' (6698172) panicked at compiler/rustc_middle/src/ty/adt.rs:462:9:
assertion failed: self.is_struct() || self.is_union()
stack backtrace:
   0: __rustc::rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::panicking::panic
   3: <rustc_hir_typeck::inline_asm::InlineAsmCtxt>::check_asm_operand_type
   4: rustc_hir_typeck::typeck_with_inspect::{closure#0}
   5: rustc_hir_typeck::typeck
      [... omitted 1 frame ...]
   6: <rustc_middle::ty::context::TyCtxt>::par_hir_body_owners::<rustc_hir_analysis::check_crate::{closure#2}>::{closure#0}
   7: rustc_hir_analysis::check_crate
   8: rustc_interface::passes::analysis
      [... omitted 1 frame ...]
   9: rustc_interface::passes::create_and_enter_global_ctxt::<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>
  10: rustc_interface::interface::run_compiler::<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: please attach the file at `/Users/jb/Library/CloudStorage/Dropbox/Desk/2503_rustc_ice/20251106 ice 보고용/rustc-ice-2025-11-07T07_18_16-4837.txt` to your bug report

query stack during panic:
#0 [typeck] type-checking `main`
#1 [analysis] running analysis passes on crate `14`
end of query stack
error: aborting due to 3 previous errors

Some errors have detailed explanations: E0084, E0517.
For more information about an error, try `rustc --explain E0084`.
```

</p>
</details>

### Notes
- ICE location: `compiler/rustc_middle/src/ty/adt.rs Line-462`
https://github.com/rust-lang/rust/blob/c90bcb9571b7aab0d8beaa2ce8a998ffaf079d38/compiler/rustc_middle/src/ty/adt.rs#L460-L464
  - In my guess, the `non_enum_variant` call comes from here:
https://github.com/rust-lang/rust/blob/c90bcb9571b7aab0d8beaa2ce8a998ffaf079d38/compiler/rustc_hir_typeck/src/inline_asm.rs#L168-L189

- Duplication Check
  - I searched "self.is_struct()" ICE ([Github issue page with query](https://github.com/rust-lang/rust/issues?q=is%3Aissue%20state%3Aclosed%20label%3AI-ICE%20%22self.is_struct()%22)), and there are no opened issues there.

--- a/compiler/rustc_hir_typeck/src/inline_asm.rs
+++ b/compiler/rustc_hir_typeck/src/inline_asm.rs
@@ -7,7 +7,7 @@ use rustc_middle::bug;
 use rustc_middle::ty::{self, Article, FloatTy, IntTy, Ty, TyCtxt, TypeVisitableExt, UintTy};
 use rustc_session::lint;
 use rustc_span::def_id::LocalDefId;
-use rustc_span::{Span, Symbol, sym};
+use rustc_span::{ErrorGuaranteed, Span, Symbol, sym};
 use rustc_target::asm::{
     InlineAsmReg, InlineAsmRegClass, InlineAsmRegOrRegClass, InlineAsmType, ModifierInfo,
 };
@@ -27,6 +27,7 @@ enum NonAsmTypeReason<'tcx> {
     InvalidElement(DefId, Ty<'tcx>),
     NotSizedPtr(Ty<'tcx>),
     EmptySIMDArray(Ty<'tcx>),
+    Tainted(ErrorGuaranteed),
 }
 
 impl<'a, 'tcx> InlineAsmCtxt<'a, 'tcx> {
@@ -94,11 +95,11 @@ impl<'a, 'tcx> InlineAsmCtxt<'a, 'tcx> {
             }
             ty::Adt(adt, args) if adt.repr().simd() => {
                 if !adt.is_struct() {
-                    self.fcx.dcx().span_delayed_bug(
+                    let guar = self.fcx.dcx().span_delayed_bug(
                         span,
                         format!("repr(simd) should only be used on structs, got {}", adt.descr()),
                     );
-                    return Err(NonAsmTypeReason::Invalid(ty));
+                    return Err(NonAsmTypeReason::Tainted(guar));
                 }
 
                 let fields = &adt.non_enum_variant().fields;
@@ -241,6 +242,9 @@ impl<'a, 'tcx> InlineAsmCtxt<'a, 'tcx> {
                     NonAsmTypeReason::EmptySIMDArray(ty) => {
                         let msg = format!("use of empty SIMD vector `{ty}`");
                         self.fcx.dcx().struct_span_err(expr.span, msg).emit();
+                    }
+                    NonAsmTypeReason::Tainted(_error_guard) => {
+                        // An error has already been reported.
                     }
                 }
                 return None;

--- a/compiler/rustc_hir_typeck/src/inline_asm.rs
+++ b/compiler/rustc_hir_typeck/src/inline_asm.rs
@@ -93,6 +93,14 @@ impl<'a, 'tcx> InlineAsmCtxt<'a, 'tcx> {
                 }
             }
             ty::Adt(adt, args) if adt.repr().simd() => {
+                if !adt.is_struct() {
+                    self.fcx.dcx().span_delayed_bug(
+                        span,
+                        format!("repr(simd) should only be used on structs, got {}", adt.descr()),
+                    );
+                    return Err(NonAsmTypeReason::Invalid(ty));
+                }
+
                 let fields = &adt.non_enum_variant().fields;
                 if fields.is_empty() {
                     return Err(NonAsmTypeReason::EmptySIMDArray(ty));

--- a/tests/ui/asm/invalid-repr-simd-on-enum-148634.rs
+++ b/tests/ui/asm/invalid-repr-simd-on-enum-148634.rs
@@ -1,3 +1,4 @@
+//@ needs-asm-support
 #![feature(repr_simd)]
 
 use std::arch::asm;
@@ -11,6 +12,5 @@ fn main() {
     unsafe {
         let mut x: Es;
         asm!("{}", out(reg) x);
-        //~^ ERROR cannot use value of type `Es` for inline assembly
     }
 }

--- a/tests/ui/asm/invalid-repr-simd-on-enum-148634.rs
+++ b/tests/ui/asm/invalid-repr-simd-on-enum-148634.rs
@@ -1,0 +1,16 @@
+#![feature(repr_simd)]
+
+use std::arch::asm;
+
+#[repr(simd)]
+//~^ ERROR attribute should be applied to a struct
+//~| ERROR unsupported representation for zero-variant enum
+enum Es {}
+
+fn main() {
+    unsafe {
+        let mut x: Es;
+        asm!("{}", out(reg) x);
+        //~^ ERROR cannot use value of type `Es` for inline assembly
+    }
+}

--- a/tests/ui/asm/invalid-repr-simd-on-enum-148634.stderr
+++ b/tests/ui/asm/invalid-repr-simd-on-enum-148634.stderr
@@ -1,5 +1,5 @@
 error[E0517]: attribute should be applied to a struct
-  --> $DIR/invalid-repr-simd-on-enum-148634.rs:5:8
+  --> $DIR/invalid-repr-simd-on-enum-148634.rs:6:8
    |
 LL | #[repr(simd)]
    |        ^^^^
@@ -8,7 +8,7 @@ LL | enum Es {}
    | ---------- not a struct
 
 error[E0084]: unsupported representation for zero-variant enum
-  --> $DIR/invalid-repr-simd-on-enum-148634.rs:5:8
+  --> $DIR/invalid-repr-simd-on-enum-148634.rs:6:8
    |
 LL | #[repr(simd)]
    |        ^^^^
@@ -16,15 +16,7 @@ LL | #[repr(simd)]
 LL | enum Es {}
    | ------- zero-variant enum
 
-error: cannot use value of type `Es` for inline assembly
-  --> $DIR/invalid-repr-simd-on-enum-148634.rs:13:29
-   |
-LL |         asm!("{}", out(reg) x);
-   |                             ^
-   |
-   = note: only integers, floats, SIMD vectors, pointers and function pointers can be used as arguments for inline assembly
-
-error: aborting due to 3 previous errors
+error: aborting due to 2 previous errors
 
 Some errors have detailed explanations: E0084, E0517.
 For more information about an error, try `rustc --explain E0084`.

--- a/tests/ui/asm/invalid-repr-simd-on-enum-148634.stderr
+++ b/tests/ui/asm/invalid-repr-simd-on-enum-148634.stderr
@@ -1,0 +1,30 @@
+error[E0517]: attribute should be applied to a struct
+  --> $DIR/invalid-repr-simd-on-enum-148634.rs:5:8
+   |
+LL | #[repr(simd)]
+   |        ^^^^
+...
+LL | enum Es {}
+   | ---------- not a struct
+
+error[E0084]: unsupported representation for zero-variant enum
+  --> $DIR/invalid-repr-simd-on-enum-148634.rs:5:8
+   |
+LL | #[repr(simd)]
+   |        ^^^^
+...
+LL | enum Es {}
+   | ------- zero-variant enum
+
+error: cannot use value of type `Es` for inline assembly
+  --> $DIR/invalid-repr-simd-on-enum-148634.rs:13:29
+   |
+LL |         asm!("{}", out(reg) x);
+   |                             ^
+   |
+   = note: only integers, floats, SIMD vectors, pointers and function pointers can be used as arguments for inline assembly
+
+error: aborting due to 3 previous errors
+
+Some errors have detailed explanations: E0084, E0517.
+For more information about an error, try `rustc --explain E0084`.

